From e32cf618d81c1bc169ec4a25fda66855d2caf933 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Tue, 19 Nov 2024 15:31:47 +0100
Subject: [PATCH] soc: st: add stm32 hog driver

This driver is used to configure the board by enabling clocks
that do not have associated drivers.

Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I620476ea60556a968defe74d3f9fd9fba02ce19c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/419941
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/soc/st/Kconfig     |  6 ++++
 drivers/soc/st/Makefile    |  1 +
 drivers/soc/st/stm32_hog.c | 64 ++++++++++++++++++++++++++++++++++++++
 3 files changed, 71 insertions(+)
 create mode 100644 drivers/soc/st/stm32_hog.c

diff --git a/drivers/soc/st/Kconfig b/drivers/soc/st/Kconfig
index 1bde2c16c21e..704a19692495 100644
--- a/drivers/soc/st/Kconfig
+++ b/drivers/soc/st/Kconfig
@@ -1,5 +1,11 @@
 if ARCH_STM32
 
+config STM32_HOG
+	tristate "STM32 hog"
+	default y
+	help
+	  Say y to enable clock(s) at the start of the Kernel.
+
 config STM32_PM_DOMAINS
 	bool "STM32 PM domains"
 	depends on MACH_STM32MP157
diff --git a/drivers/soc/st/Makefile b/drivers/soc/st/Makefile
index 6c71607f6c89..11f056bfc80a 100644
--- a/drivers/soc/st/Makefile
+++ b/drivers/soc/st/Makefile
@@ -1,3 +1,4 @@
+obj-$(CONFIG_STM32_HOG) += stm32_hog.o
 obj-$(CONFIG_STM32_PM_DOMAINS) += stm32_pm_domain.o
 obj-$(CONFIG_STM32_RISAB) += stm32_risab.o
 obj-$(CONFIG_STM32_RISAF) += stm32_risaf.o
diff --git a/drivers/soc/st/stm32_hog.c b/drivers/soc/st/stm32_hog.c
new file mode 100644
index 000000000000..c8fbf8d5e57e
--- /dev/null
+++ b/drivers/soc/st/stm32_hog.c
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2024, STMicroelectronics - All Rights Reserved
+ */
+
+#include <linux/clk.h>
+#include "linux/device.h"
+#include <linux/of_address.h>
+#include <linux/platform_device.h>
+
+struct stm32_hog_pdata {
+	struct clk_bulk_data *clks;
+	int num_clks;
+};
+
+static void stm32_hog_remove(struct platform_device *pdev)
+{
+	struct stm32_hog_pdata *priv = platform_get_drvdata(pdev);
+
+	clk_bulk_disable_unprepare(priv->num_clks, priv->clks);
+}
+
+static int stm32_hog_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct stm32_hog_pdata *priv;
+	int ret;
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	priv->num_clks = devm_clk_bulk_get_all(dev, &priv->clks);
+	if (priv->num_clks < 1)
+		return -ENODEV;
+
+	ret = clk_bulk_prepare_enable(priv->num_clks, priv->clks);
+	if (ret) {
+		dev_err(dev, "failed to enable bulk clks %d\n", ret);
+		return ret;
+	}
+
+	platform_set_drvdata(pdev, priv);
+
+	return 0;
+}
+
+static const struct of_device_id stm32_hog_match[] = {
+	{ .compatible = "st,stm32-hog", },
+	{}
+};
+MODULE_DEVICE_TABLE(of, stm32_hog_match);
+
+static struct platform_driver stm32_hog_driver = {
+	.probe  = stm32_hog_probe,
+	.remove_new = stm32_hog_remove,
+	.driver = {
+		.name = "stm32-hog",
+		.of_match_table = stm32_hog_match,
+	},
+};
+module_platform_driver(stm32_hog_driver);
+MODULE_AUTHOR("Gabriel Fernandez <gabriel.fernandez@foss.st.com>");
+MODULE_LICENSE("GPL");
-- 
2.50.1

